Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp

```diff
@@ -502,88 +605,253 @@ void ProgramShaderCache::Init()
+                                                    u32 value_size)
+{
+  entry->in_cache = true;
+  entry->pending = false;
+  entry->shader.glprogid = CreateProgramFromBinary(value, value_size);
+  if (entry->shader.glprogid == 0)
+    return false;
 
-      std::vector<u8> data(binary_size + sizeof(GLenum));
-      u8* binary = &data[sizeof(GLenum)];
-      GLenum* prog_format = (GLenum*)&data[0];
-      glGetProgramBinary(entry.second.shader.glprogid, binary_size, nullptr, prog_format, binary);
-      if (glGetError() != GL_NO_ERROR)
-      {
-        continue;
-      }
+  entry->shader.SetProgramVariables();
+  return true;
+}
 
-      g_program_disk_cache.Append(entry.first, &data[0], binary_size + sizeof(GLenum));
-    }
+void ProgramShaderCache::LoadProgramBinaries()
+{
+  GLint Supported;
+  glGetIntegerv(GL_NUM_PROGRAM_BINARY_FORMATS, &Supported);
+  if (!Supported)
+  {
+    ERROR_LOG(VIDEO, "GL_ARB_get_program_binary is supported, but no binary format is known. So "
+                     "disable shader cache.");
+    g_ogl_config.bSupportsGLSLCache = false;
+  }
+  else
+  {
+    // Load game-specific shaders.
+    std::string cache_filename =
+        GetDiskShaderCacheFileName(APIType::OpenGL, "ProgramBinaries", true, true);
+    ProgramShaderCacheInserter<SHADERUID> inserter(pshaders);
+    s_program_disk_cache.OpenAndRead(cache_filename, inserter);
+
+    // Load global ubershaders.
+    cache_filename =
+        GetDiskShaderCacheFileName(APIType::OpenGL, "UberProgramBinaries", false, true);
+    ProgramShaderCacheInserter<UBERSHADERUID> uber_inserter(ubershaders);
+    s_uber_program_disk_cache.OpenAndRead(cache_filename, uber_inserter);
+  }
+  SETSTAT(stats.numPixelShadersAlive, pshaders.size());
+}
 
-    g_program_disk_cache.Sync();
-    g_program_disk_cache.Close();
+bool ProgramShaderCache::GetProgramBinary(const PCacheEntry& entry, std::vector<u8>& data)
+{
+  // Clear any prior error code
+  glGetError();
+
+  GLint link_status = GL_FALSE, delete_status = GL_TRUE, binary_size = 0;
+  glGetProgramiv(entry.shader.glprogid, GL_LINK_STATUS, &link_status);
+  glGetProgramiv(entry.shader.glprogid, GL_DELETE_STATUS, &delete_status);
+  glGetProgramiv(entry.shader.glprogid, GL_PROGRAM_BINARY_LENGTH, &binary_size);
+  if (glGetError() != GL_NO_ERROR || link_status == GL_FALSE || delete_status == GL_TRUE ||
+      binary_size == 0)
+  {
+    return false;
   }
 
-  glUseProgram(0);
+  data.resize(binary_size + sizeof(GLenum));
+
+  GLsizei length = binary_size;
+  GLenum prog_format;
+  glGetProgramBinary(entry.shader.glprogid, binary_size, &length, &prog_format,
+                     &data[sizeof(GLenum)]);
+  if (glGetError() != GL_NO_ERROR)
+    return false;
+
+  std::memcpy(&data[0], &prog_format, sizeof(prog_format));
+  return true;
+}
+
+void ProgramShaderCache::SaveProgramBinaries()
+{
+  std::vector<u8> binary_data;
 
   for (auto& entry : pshaders)
```

Comment:
This near exact code makes me again wonder if we could combine the caches into one.  Probably more work than it's worth.